File: app/src/main/java/com/kieran/winnipegbus/Activities/StopTimesActivity.java
package com.kieran.winnipegbus.Activities;

import android.app.AlertDialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.graphics.drawable.Drawable;
import android.hardware.Sensor;
import android.hardware.SensorManager;
import android.os.AsyncTask;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.v4.widget.SwipeRefreshLayout;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ListView;
import android.widget.TextView;

import com.kieran.winnipegbus.Adapters.StopTimeAdapter;
import com.kieran.winnipegbus.LoadXMLAsyncTask;
import com.kieran.winnipegbus.NotificationService;
import com.kieran.winnipegbus.R;
import com.kieran.winnipegbus.ShakeDetector;
import com.kieran.winnipegbusbackend.BusUtilities;
import com.kieran.winnipegbusbackend.FavouriteStop;
import com.kieran.winnipegbusbackend.FavouriteStopsList;
import com.kieran.winnipegbusbackend.LoadResult;
import com.kieran.winnipegbusbackend.RouteSchedule;
import com.kieran.winnipegbusbackend.ScheduledStop;
import com.kieran.winnipegbusbackend.Stop;
import com.kieran.winnipegbusbackend.StopSchedule;
import com.kieran.winnipegbusbackend.StopTime;

import org.w3c.dom.Document;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class StopTimesActivity extends BaseActivity implements SwipeRefreshLayout.OnRefreshListener, AdapterView.OnItemClickListener, ShakeDetector.OnShakeListener, AdapterView.OnItemLongClickListener {

    private static final String FILTER_POSITIVE = "Done";
    public static final String STOP = "stop";
    private static final String UPDATED_STRING = "Updated %s";
    private static final String ACTIONBAR_TEXT = "Stop %d";
    private static final String DELETE_THIS_FAVOURITE = "Delete this Favourite?";
    private static final String DIALOG_YES = "Yes";
    private static final String DIALOG_NO = "No";
    private static final String CREATE_NOTIFICATION_FOR_BUS = "Create a notification for this bus?";
    private StopSchedule stopSchedule; //TODO: static?

    private int stopNumber;
    private String stopName;
    private boolean loading = false;
    private List<ScheduledStop> stops = new ArrayList<>();
    private StopTimeAdapter adapter;
    private List<Integer> routeNumberFilter = new ArrayList<>();
    private TextView title;
    private boolean[] selectedRoutes;
    private List<RouteSchedule> routeFilterRoutes = new ArrayList<>();
    private AsyncTask loadStopTimesTask;
    private SensorManager sensorManager;
    private Sensor accelerometer;
    private ShakeDetector shakeDetector;
    private boolean hasFilterChanged;
    private SwipeRefreshLayout swipeRefreshLayout;
    private TextView lastUpdated;

    @Override
    protected void onNewIntent(Intent intent) {
        int newStopNumber = ((Stop) intent.getSerializableExtra(STOP)).getNumber();
        if (newStopNumber != stopNumber) {
            stopNumber = newStopNumber;
            setIntent(intent);
            recreate();
        }
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_stop_times);
        adViewResId = R.id.stopTimesAdView;

        //stopSchedule = null; //TODO remove dis
        // test
        FavouriteStopsList.loadFavourites();

        ListView listView = (ListView) findViewById(R.id.stop_times_listview);

        swipeRefreshLayout = (SwipeRefreshLayout) findViewById(R.id.stop_times_swipe_refresh);
        swipeRefreshLayout.setOnRefreshListener(this);
        swipeRefreshLayout.setColorSchemeResources(R.color.rt_blue, R.color.rt_red);

        listView.setOnItemClickListener(this);
        //listView.setOnItemLongClickListener(this); //TODO disable to remove notification system

        adapter = new StopTimeAdapter(this, R.layout.listview_stop_times_row, stops);
        listView.addHeaderView(getLayoutInflater().inflate(R.layout.listview_stop_times_header, null));
        listView.setAdapter(adapter);

        Intent intent = getIntent();

        if (title == null) {
            title = ((TextView) findViewById(R.id.listView_stop_times_header_text));
            lastUpdated = (TextView) findViewById(R.id.stop_times_header_last_updated);
        }

        Stop stop = (Stop) intent.getSerializableExtra(STOP);
        stopName = stop.getName();
        title.setText(stopName);
        stopNumber = stop.getNumber();

        setTitle(String.format(ACTIONBAR_TEXT, stopNumber));

        initializeAdsIfEnabled();
        createShakeListener();
    }

    public void createShakeListener() {
        PackageManager PM = getPackageManager();

        if (PM.hasSystemFeature(PackageManager.FEATURE_SENSOR_ACCELEROMETER)) {
            sensorManager = (SensorManager) getSystemService(SENSOR_SERVICE);
            accelerometer = sensorManager.getDefaultSensor(Sensor.TYPE_ACCELEROMETER);
            shakeDetector = new ShakeDetector(this);
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        loading = false;
        if (loadStopTimesTask != null)
            loadStopTimesTask.cancel(true);
    }

    @Override
    public void onResume() {
        super.onResume();
        sensorManager.registerListener(shakeDetector, accelerometer, SensorManager.SENSOR_DELAY_UI);
    }

    @Override
    public void onPause() {
        super.onPause();
        loading = false;
        sensorManager.unregisterListener(shakeDetector);
    }

    private void openAdditionalInfo(int position) {
        Intent intent;
        if (stopSchedule != null) {
            if (position == 0) {
                intent = new Intent(this, StopInfoActivity.class);
                intent.putExtra(StopInfoActivity.STOP_FEATURES, stopSchedule.createStopFeatures());
            } else {
                intent = new Intent(this, ScheduledStopInfoActivity.class);
                intent.putExtra(ScheduledStopInfoActivity.STOP, stops.get(position - 1));
            }
            startActivity(intent);
        }

    }

    private void getTimes() {
        String urlPath = BusUtilities.generateStopNumberURL(stopNumber, routeNumberFilter, null, getScheduleEndTime());

        loadStopTimesTask = new LoadStopTimes().execute(urlPath);
    }

    private StopTime getScheduleEndTime() {
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(this);
        StopTime endTime = new StopTime(System.currentTimeMillis());
        endTime.increaseHour(Integer.parseInt(prefs.getString(getString(R.string.pref_schedule_load_interval), "2")));

        return endTime;
    }

    @Override
    public boolean onCreateOptionsMenu(final Menu menu) {
        getMenuInflater().inflate(R.menu.menu_stop_times, menu);

        menu.findItem(R.id.add_to_favourites_button).setIcon(getFavouritesButtonDrawable(FavouriteStopsList.contains(stopNumber)));
        onOptionsItemSelected(menu.findItem(R.id.action_refresh)); //manually click the refresh button, this is the only way the swipe refresh loading spinner works correctly on initial load. Not happy with this but it was the only way I could get it to work
        return true;
    }

    @SuppressWarnings("deprecation")
    private Drawable getFavouritesButtonDrawable(boolean isFavoured) {
        int themeId = super.getThemeResId();
        int drawableId;

        if (themeId == R.style.Dark || themeId == R.style.Rt)
            if (isFavoured)
                drawableId = R.drawable.ic_favourite_stops_dark;
            else
                drawableId = R.drawable.ic_add_to_favourites_dark;
        else if (isFavoured)
            drawableId = R.drawable.ic_favourite_stops_light;
        else
            drawableId = R.drawable.ic_add_to_favourites_light;

        return getResources().getDrawable(drawableId);
    }

    private void refresh() {
        if (!loading) {
            if (isOnline()) {
                loading = true;

                adapter.loadTimeSetting();

                getTimes();
            } else {
                showLongToaster(getText(R.string.network_error).toString());
            }
        }
        swipeRefreshLayout.setRefreshing(loading);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_refresh:
                refresh();
                return true;
            case R.id.add_to_favourites_button:
                handleFavouritesClick(item);
                return true;
            case R.id.action_filter:
                if (!loading && stopSchedule != null)
                    openFilterWindow();
                else
                    showLongToaster(getString(R.string.wait_for_load));
                return true;

        }
        return super.onOptionsItemSelected(item);
    }

    private void handleFavouritesClick(final MenuItem item) {
        if (FavouriteStopsList.contains(stopNumber)) {
            AlertDialog.Builder alertDialog = new AlertDialog.Builder(this);
            alertDialog.setMessage(DELETE_THIS_FAVOURITE);

            alertDialog.setPositiveButton(DIALOG_YES, new DialogInterface.OnClickListener() {
                public void onClick(DialogInterface dialogInterface, int which) {
                    FavouriteStopsList.removeFromFavourites(stopNumber);
                    item.setIcon(getFavouritesButtonDrawable(false));
                }
            });

            alertDialog.setNegativeButton(DIALOG_NO, null);
            alertDialog.create().show();
        } else if (stopName != null && !stopName.equals("")) {
            FavouriteStopsList.addToFavourites(new FavouriteStop(stopName, stopNumber));
            item.setIcon(getFavouritesButtonDrawable(true));
        } else {
            showLongToaster(getString(R.string.wait_for_load));
        }

    }

    private void openFilterWindow() {
        if (stopSchedule != null) {
            hasFilterChanged = false;
            AlertDialog.Builder filterDialog = new AlertDialog.Builder(this);
            filterDialog.setTitle(R.string.filter_dialog_title);

            if (routeFilterRoutes.size() == 0)
                getFilterRoutes();

            if (routeFilterRoutes.size() < stopSchedule.getRouteList().size()) {
                routeFilterRoutes.clear();
                getFilterRoutes();
            }

            CharSequence charSequence[] = new CharSequence[routeFilterRoutes.size()];

            if (selectedRoutes == null)
                selectedRoutes = new boolean[routeFilterRoutes.size()];

            for (int i = 0; i < charSequence.length; i++)
                charSequence[i] = routeFilterRoutes.get(i).toString();

            filterDialog.setMultiChoiceItems(charSequence, selectedRoutes, new AlertDialog.OnMultiChoiceClickListener() {
                @Override
                public void onClick(DialogInterface dialog, int which, boolean isChecked) {
                    hasFilterChanged = true;
                    int routeNumber = routeFilterRoutes.get(which).getRouteNumber();
                    selectedRoutes[which] = isChecked;
                    if (isChecked)
                        routeNumberFilter.add(routeNumber);
                    else
                        routeNumberFilter.remove(Integer.valueOf(routeNumber));
                }
            });

            filterDialog.setPositiveButton(FILTER_POSITIVE, new DialogInterface.OnClickListener() {
                @Override
                public void onClick(DialogInterface dialog, int which) {
                    if (hasFilterChanged) {
                        refresh();
                    }
                }
            });

            filterDialog.create().show();
        } else {
            showLongToaster(getString(R.string.wait_for_load));
        }
    }

    public void getFilterRoutes() {
        for (RouteSchedule routeSchedule : stopSchedule.getRouteList())
            routeFilterRoutes.add(new RouteSchedule(routeSchedule));

        Collections.sort(routeFilterRoutes);
    }

    @Override
    public void onRefresh() {
        refresh();
    }

    @Override
    public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
        openAdditionalInfo(position);
    }

    @Override
    public void onShake() {
        refresh();
    }

    @Override
    public boolean onItemLongClick(AdapterView<?> parent, View view, final int position, long id) {
        if (position != 0) {
            AlertDialog.Builder alertDialog = new AlertDialog.Builder(this);
            alertDialog.setMessage(CREATE_NOTIFICATION_FOR_BUS);
            alertDialog.setPositiveButton(DIALOG_YES, new DialogInterface.OnClickListener() {
                public void onClick(DialogInterface dialogInterface, int which) {
                    ScheduledStop selectedStop = stops.get(position - 1);
                    NotificationService.createNotification(stopNumber, selectedStop.getRouteNumber(), selectedStop.getKey(), selectedStop.getRouteVariantName(), stopName, selectedStop.getEstimatedDepartureTime(), getApplicationContext(), selectedStop.getCoverageType());
                }
            });

            alertDialog.setNegativeButton(DIALOG_NO, null);
            alertDialog.create().show();
            return true;
        }
        return false;
    }

    private class LoadStopTimes extends LoadXMLAsyncTask {
        @Override
        protected LoadResult doInBackground(String... urls) {
            LoadResult result = super.doInBackground(urls);

            if (loading && result.getResult() != null) {
                if (stopSchedule == null) {
                    stopSchedule = new StopSchedule((Document) result.getResult(), stopNumber);
                    stopName = stopSchedule.getName();
                } else {
                    stopSchedule.refresh((Document) result.getResult());
                }
                stops.clear();
                stops.addAll(stopSchedule.getScheduledStopsSorted());
            }
            return result;
        }

        @Override
        protected void onPostExecute(LoadResult result) {
            if (result.getException() != null && loading) {
                showLongToaster(getText(R.string.network_error).toString());
                if (stopSchedule == null)
                    title.setText(R.string.network_error);
            } else if (stops.size() == 0 && loading) {
                showLongToaster(getText(R.string.no_results_found).toString());
                title.setText(R.string.no_results_found);
            } else {
                title.setText(stopName);
            }

            adapter.notifyDataSetChanged();

            lastUpdated.setText(String.format(UPDATED_STRING, new StopTime(System.currentTimeMillis()).toFormattedString(null, getTimeSetting())));
            swipeRefreshLayout.setRefreshing(false);
            loading = false;
        }
    }
}

File: app/src/main/java/com/kieran/winnipegbus/Activities/StopTimesActivity.kt
package com.kieran.winnipegbus.Activities

import android.app.AlertDialog
import android.content.DialogInterface
import android.content.Intent
import android.content.SharedPreferences
import android.content.pm.PackageManager
import android.graphics.drawable.Drawable
import android.hardware.Sensor
import android.hardware.SensorManager
import android.os.AsyncTask
import android.os.Bundle
import android.preference.PreferenceManager
import android.support.v4.widget.SwipeRefreshLayout
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.AdapterView
import android.widget.ListView
import android.widget.TextView

import com.kieran.winnipegbus.Adapters.StopTimeAdapter
import com.kieran.winnipegbus.R
import com.kieran.winnipegbus.ShakeDetector
import com.kieran.winnipegbus.Views.StyledSwipeRefresh
import com.kieran.winnipegbusbackend.FavouriteStop
import com.kieran.winnipegbusbackend.FavouriteStopsList
import com.kieran.winnipegbusbackend.LoadResult
import com.kieran.winnipegbusbackend.Route
import com.kieran.winnipegbusbackend.RouteSchedule
import com.kieran.winnipegbusbackend.ScheduledStop
import com.kieran.winnipegbusbackend.Stop
import com.kieran.winnipegbusbackend.StopSchedule
import com.kieran.winnipegbusbackend.StopTime
import com.kieran.winnipegbusbackend.TransitApiManager

import org.json.JSONObject

import java.io.IOException
import java.util.ArrayList
import java.util.Collections
import java.util.Locale

class StopTimesActivity : BaseActivity(), SwipeRefreshLayout.OnRefreshListener, AdapterView.OnItemClickListener, ShakeDetector.OnShakeListener {
    private var stopSchedule: StopSchedule? = null

    private var stopNumber: Int = 0
    private var stopName: String? = null
    private var loading = false
    private val stops = ArrayList<ScheduledStop>()
    private var adapter: StopTimeAdapter? = null
    private val routeNumberFilter = ArrayList<Int>()
    private var title: TextView? = null
    private var selectedRoutes: BooleanArray? = null
    private val routeFilterRoutes = ArrayList<Route>()
    private var loadStopTimesTask: AsyncTask<*, *, *>? = null
    private var sensorManager: SensorManager? = null
    private var accelerometer: Sensor? = null
    private var shakeDetector: ShakeDetector? = null
    private var hasFilterChanged: Boolean = false
    private var swipeRefreshLayout: StyledSwipeRefresh? = null
    private var lastUpdated: TextView? = null

    private val scheduleEndTime: StopTime
        get() {
            val prefs = PreferenceManager.getDefaultSharedPreferences(this)
            val endTime = StopTime()
            endTime.increaseHour(Integer.parseInt(prefs.getString(getString(R.string.pref_schedule_load_interval), "2")))

            return endTime
        }

    override fun onNewIntent(intent: Intent) {
        val newStopNumber = (intent.getSerializableExtra(STOP) as Stop).number
        if (newStopNumber != stopNumber) {
            stopNumber = newStopNumber
            setIntent(intent)
            recreate()
        }
    }

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_stop_times)
        adViewResId = R.id.stopTimesAdView

        FavouriteStopsList.loadFavourites()

        val listView = findViewById<ListView>(R.id.stop_times_listview)

        swipeRefreshLayout = findViewById(R.id.stop_times_swipe_refresh)
        swipeRefreshLayout!!.setOnRefreshListener(this)


        listView.onItemClickListener = this

        adapter = StopTimeAdapter(this, R.layout.listview_stop_times_row, stops)
        listView.addHeaderView(layoutInflater.inflate(R.layout.listview_stop_times_header, null))
        listView.adapter = adapter

        val intent = intent

        if (title == null) {
            title = findViewById(R.id.listView_stop_times_header_text)
            lastUpdated = findViewById(R.id.stop_times_header_last_updated)
        }

        val stop = intent.getSerializableExtra(STOP) as Stop
        stopName = stop.name
        title!!.text = stopName
        stopNumber = stop.number

        setTitle(String.format(Locale.CANADA, ACTIONBAR_TEXT, stopNumber))

        initializeAdsIfEnabled()
        createShakeListener()
        loading = false
    }

    fun createShakeListener() {
        val PM = packageManager

        if (PM.hasSystemFeature(PackageManager.FEATURE_SENSOR_ACCELEROMETER)) {
            sensorManager = getSystemService(Context.SENSOR_SERVICE) as SensorManager
            accelerometer = sensorManager!!.getDefaultSensor(Sensor.TYPE_ACCELEROMETER)
            shakeDetector = ShakeDetector(this)
        }
    }

    public override fun onDestroy() {
        super.onDestroy()
        loading = false
        if (loadStopTimesTask != null)
            loadStopTimesTask!!.cancel(true)
    }

    public override fun onResume() {
        super.onResume()

        if (sensorManager != null)
            sensorManager!!.registerListener(shakeDetector, accelerometer, SensorManager.SENSOR_DELAY_UI)
    }

    public override fun onPause() {
        super.onPause()
        loading = false

        if (sensorManager != null)
            sensorManager!!.unregisterListener(shakeDetector)
    }

    private fun openAdditionalInfo(position: Int) {
        val intent: Intent
        if (stopSchedule != null) {
            if (position == 0) {
                intent = Intent(this, StopInfoActivity::class.java)
                intent.putExtra(StopInfoActivity.STOP_FEATURES, stopSchedule!!.createStopFeatures())
            } else {
                intent = Intent(this, ScheduledStopInfoActivity::class.java)
                intent.putExtra(ScheduledStopInfoActivity.STOP_EXTRA, stops[position - 1])
            }
            startActivity(intent)
        }
    }

    private fun getTimes() {
        val urlPath = TransitApiManager.generateStopNumberURL(stopNumber, routeNumberFilter, null, scheduleEndTime)

        loadStopTimesTask = LoadStopTimes().execute(urlPath)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_stop_times, menu)

        menu.findItem(R.id.add_to_favourites_button).icon = getFavouritesButtonDrawable(FavouriteStopsList.contains(stopNumber))
        onOptionsItemSelected(menu.findItem(R.id.action_refresh)) //manually click the refresh button, this is the only way the swipe refresh loading spinner works correctly on initial load. Not happy with this but it was the only way I could get it to work
        return true
    }

    private fun getFavouritesButtonDrawable(isFavoured: Boolean): Drawable {
        val themeId = themeResId
        val drawableId: Int

        if (isFavoured)
            drawableId = if (themeId == R.style.Light) R.drawable.ic_favourite_stops_light else R.drawable.ic_favourite_stops_dark
        else
            drawableId = if (themeId == R.style.Light) R.drawable.ic_add_to_favourites_light else R.drawable.ic_add_to_favourites_dark

        return resources.getDrawable(drawableId)
    }

    private fun refresh() {
        if (!loading) {
            if (isOnline) {
                loading = true

                adapter!!.loadTimeSetting()

                getTimes()
            } else {
                showLongToaster(R.string.network_error)
            }
        }
        swipeRefreshLayout!!.isRefreshing = loading
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_refresh -> {
                refresh()
                return true
            }
            R.id.add_to_favourites_button -> {
                handleFavouritesClick(item)
                return true
            }
            R.id.action_filter -> {
                if (!loading && stopSchedule != null)
                    openFilterWindow()
                else
                    showLongToaster(R.string.wait_for_load)
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

    private fun handleFavouritesClick(item: MenuItem) {
        if (FavouriteStopsList.contains(stopNumber)) {
            openDeleteFavouriteDialog(item)
        } else if (stopName != null && stopName != "") {
            FavouriteStopsList.addToFavourites(FavouriteStop(stopName, stopNumber))
            item.icon = getFavouritesButtonDrawable(true)
        } else {
            showLongToaster(R.string.wait_for_load)
        }
    }

    private fun openDeleteFavouriteDialog(item: MenuItem) {
        val alertDialog = AlertDialog.Builder(this)
        alertDialog.setMessage(DELETE_THIS_FAVOURITE)

        alertDialog.setPositiveButton(DIALOG_YES) { dialogInterface, which ->
            FavouriteStopsList.remove(stopNumber)
            item.icon = getFavouritesButtonDrawable(false)
        }

        alertDialog.setNegativeButton(DIALOG_NO, null)
        alertDialog.create().show()
    }

    private fun openFilterWindow() {
        if (stopSchedule != null) {
            hasFilterChanged = false
            val filterDialog = AlertDialog.Builder(this)
            filterDialog.setTitle(R.string.filter_dialog_title)

            if (routeFilterRoutes.size == 0)
                getFilterRoutes()

            if (routeFilterRoutes.size < stopSchedule!!.routeList.size) {
                routeFilterRoutes.clear()
                getFilterRoutes()
            }

            val charSequence = arrayOfNulls<CharSequence>(routeFilterRoutes.size)

            if (selectedRoutes == null)
                selectedRoutes = BooleanArray(routeFilterRoutes.size)

            for (i in charSequence.indices)
                charSequence[i] = routeFilterRoutes[i].toString()

            filterDialog.setMultiChoiceItems(charSequence, selectedRoutes, object : AlertDialog.OnMultiChoiceClickListener {
                override fun onClick(dialog: DialogInterface, which: Int, isChecked: Boolean) {
                    hasFilterChanged = true
                    val routeNumber = routeFilterRoutes[which].routeNumber
                    selectedRoutes[which] = isChecked
                    if (isChecked)
                        routeNumberFilter.add(routeNumber)
                    else
                        routeNumberFilter.remove(Integer.valueOf(routeNumber))
                }
            })

            filterDialog.setPositiveButton(FILTER_POSITIVE) { dialog, which ->
                if (hasFilterChanged) {
                    refresh()
                }
            }

            filterDialog.create().show()
        } else {
            showLongToaster(R.string.wait_for_load)
        }
    }

    fun getFilterRoutes() {
        for (routeSchedule in stopSchedule!!.routeList)
            routeFilterRoutes.add(Route(routeSchedule))

        Collections.sort(routeFilterRoutes)
    }

    override fun onRefresh() {
        refresh()
    }

    override fun onItemClick(parent: AdapterView<*>, view: View, position: Int, id: Long) {
        openAdditionalInfo(position)
    }

    override fun onShake() {
        refresh()
    }

    private inner class LoadStopTimes : AsyncTask<String, Void, LoadResult<JSONObject>>() {
        override fun doInBackground(vararg urls: String): LoadResult<JSONObject> {
            val result: LoadResult<JSONObject>

            result = TransitApiManager.getJson(urls[0])

            if (loading && result.result != null) {
                if (stopSchedule == null) {
                    stopSchedule = StopSchedule(result.result, stopNumber)

                    stopName = stopSchedule!!.name
                } else {
                    stopSchedule!!.refresh(result.result)
                }
            }

            return result
        }

        protected override fun onPostExecute(result: LoadResult<*>) {
            if (loading && result.result != null) {
                stops.clear()
                stops.addAll(stopSchedule!!.scheduledStopsSorted)
            }

            if (result.exception != null && loading) {
                handleException(result.exception)
                if (stopSchedule == null && result.exception is IOException)
                    title!!.setText(R.string.network_error)
            } else if (stops.size == 0 && loading) {
                showLongToaster(R.string.no_results_found)
                title!!.setText(R.string.no_results_found)
            } else {
                title!!.text = stopName
            }

            adapter!!.notifyDataSetChanged()

            lastUpdated!!.text = String.format(UPDATED_STRING, StopTime().toFormattedString(null, timeSetting))
            swipeRefreshLayout!!.isRefreshing = false
            loading = false
        }
    }

    companion object {

        private val FILTER_POSITIVE = "Done"
        val STOP = "stop"
        private val UPDATED_STRING = "Updated %s"
        private val ACTIONBAR_TEXT = "Stop %d"
        private val DELETE_THIS_FAVOURITE = "Delete this Favourite?"
        private val DIALOG_YES = "Yes"
        private val DIALOG_NO = "No"
        private val CREATE_NOTIFICATION_FOR_BUS = "Create a notification for this bus?"
    }
}